src/layout/layout.rs

```diff
@@ -602,13 +656,30 @@ impl Layout {
 }
 
 impl Element {
-    fn new() -> Element {
-        Element {
+    #[allow(dead_code)]
+    fn new() -> Self {
+        Self {
             start: Variable::new(),
             end: Variable::new(),
         }
     }
 
+    fn constrain(
+        solver: &mut Solver,
+        (area_start, area_end): (f64, f64),
+    ) -> Result<Self, AddConstraintError> {
+        let e = Element {
+            start: Variable::new(),
+            end: Variable::new(),
+        };
+        solver.add_constraints(&[
+            e.start | GE(REQUIRED) | area_start,
+            e.end | LE(REQUIRED) | area_end,
+            e.start | LE(REQUIRED) | e.end,
```

Comment:
Is this constraint necessary?

Reply:
I think it is required because it is a base assumption about how we measure width. 

```
    fn size(&self) -> Expression {
        self.end - self.start
    }
```

without this constraint, the solver may find a solution where width is negative. This is probably more true now with the `Proportional` constraints and the `Flex` layouts.

Reply:
But if area_start < area_end, and e.start >= area_start, and e.end <= area.end, I'm not sure how e.start > e.end unless math is broken

Reply:
If the solver finds this `Rect` for whatever reason:

```
area_start == 0.0
e.end == 5.0
e.start == 8.0
area_end == 10.0
```

- [x] `area_start < area_end`, 0 < 10
- [x] `e.start >= area_start`, 8 >= 0
- [x] `e.end` <= area_end`, 5 <= 10

We need this constraint to ensure that this doesn't happen:

- [ ] `e.start <= e.end`, 8 >= 5

Reply:
`area_start` and `area_end` are the total area start and end. Maybe you are thinking it is the area of that particular rect?

Reply:
oh yeah, my bad.